### PR TITLE
feat: display subscribed data

### PR DESCRIPTION
**Pull request by mattydono**
Added ability for user to display 5-day forecast of data.

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -10,13 +10,35 @@ class App extends Component {
...
+                    {Object.entries(this.state.subscribedCities).map(([city, data]) => this.renderCityData(city, data))}
             </div>
         );
     }
 
+    onPrint = () => {
+        console.log(this.state.subscribedCities);
+    };
+
+    renderCityData = (city, data) => {
```
Comment (Sharakai) [inline src/App.js]: 👍 Nice function name.
The multiple `return` keywords are redundant, though:

```suggestion
    renderCityData = (city, data) => data.list.map(datum => (
      <div>{city}
        <div key={city + 'Date'}>Date: {datum.dt_txt}</div>
        <div key={city + 'temp'}>Temp: {(datum.main.temp - 273.15).toFixed(1)}ºC</div>
        <div key={city + 'humidity'}>Humidity: {datum.main.humidity}%</div>
        <div key={city + 'description'}>{datum.weather[0].description.toLocaleUpperCase()}</div>
        <div key={city + 'wind'}>Wind Speed: {(datum.wind.speed * 2.236936).toFixed(1)}mph</div>
      </div>
    );
```

File: src/App.js [hunk 2]
```diff
@@ -10,13 +10,35 @@ class App extends Component {
...
+        return data.list.map( datum => {
+            return (
+                <div>{city}
```
Comment (Sharakai) [inline src/App.js]: Why is this `{city}` left alone outside of a `<span>` or some sort of container?

File: src/App.js [hunk 3]
```diff
@@ -10,13 +10,35 @@ class App extends Component {
...
+                    <div key={city + 'Date'}>Date: {datum.dt_txt}</div>
+                    <div key={city + 'temp'}>Temp: {(datum.main.temp - 273.15).toFixed(1)}ºC</div>
+                    <div key={city + 'humidity'}>Humidity: {datum.main.humidity}%</div>
+                    <div key={city + 'description'}>{datum.weather[0].description.toLocaleUpperCase()}</div>
```
Comment (Sharakai) [inline src/App.js]: Does `datum.weather[0]` always exist? This will throw an undefined exception if it doesn't.

File: src/App.js [hunk 4]
```diff
@@ -31,14 +53,16 @@ class App extends Component {
 
     onCitySubscribe = () => {
         const citySubscribe = this.validateCity(cities);
```
Comment (Sharakai) [inline src/App.js]: Why do you need to pass `cities` into this function, when it's accessible outside the class?
This `validateCity` function is much better as a pure function:

```javascript
const isValidCity = selectedCityName => 
  cities.some(city => city.name === selectedCityName);
```

Comment (Sharakai) [reply src/App.js]: In fact, `validateCity` isn't simply validating it - It is also parsing it into a query string. This should **definitely** be split into 2 functions. One being `isValidCity` (above), and the other being `parseCityRequest`:
```javascript
const parseCityRequest = selectedCityName => {
  const city = cities.find(city => city.name === selectedCityName);
  if (city !== undefined) {
    return `${city.name},${city.country}`
  }
  return undefined;
}
```

File: src/App.js [hunk 5]
```diff
@@ -10,13 +10,35 @@ class App extends Component {
 
     render() {
         return (
-            <div className="App">
-                <input placeholder="city,country code" onChange={this.onCityChange}/>
-                <button onClick={this.onCitySubscribe}>Subscribe</button>
+            <div>
+                <div className="App">
```
Comment (Sharakai) [inline src/App.js]: Why is the `className="App"` at this level and not the `<div>` above?

Comment (Sharakai) [reply src/App.js]: Also, this block of `<input>` and `<button>`s should be a separate component, in a separate file

File: src/App.js [hunk 6]
```diff
@@ -10,13 +10,35 @@ class App extends Component {
...
+                    <input placeholder="city,country code" onChange={this.onCityChange}/>
+                    <button onClick={this.onCitySubscribe}>Subscribe</button>
+                    <button onClick={this.onPrint}>Print</button>
+                </div>
+                    {Object.entries(this.state.subscribedCities).map(([city, data]) => this.renderCityData(city, data))}
```
Comment (Sharakai) [inline src/App.js]: Lambda's in JSX are generally frowned upon, so it'd be better to change `renderCityData` to take the `[city, data]` tuple, and use:
```javascript
{Object.entries(this.state.subscribedCities).map(this.renderCityData)}
```

EDIT: In fact, why use `Object.entries` at all? The values of the object contain the `city.name` property, so we don't actually need the keys at all. This simplifies it all to:

```javascript
Object.values(this.state.subscribedCities).map(this.renderCityData)}

...

renderCityData = data => {
  const cityName = data.city.name;
  return data.list.map(datum => ...
}
```

Comment (Sharakai) [reply src/App.js]: Also, this should preferably be in a different component, in a separate file, focused solely on displaying and managing the data

File: src/App.js [hunk 7]
```diff
@@ -10,13 +10,35 @@ class App extends Component {
 
     render() {
         return (
-            <div className="App">
-                <input placeholder="city,country code" onChange={this.onCityChange}/>
-                <button onClick={this.onCitySubscribe}>Subscribe</button>
+            <div>
+                <div className="App">
+                    <input placeholder="city,country code" onChange={this.onCityChange}/>
```
Comment (Sharakai) [inline src/App.js]: You're missing the `value={this.state.selectedCity}` prop:
```suggestion
                    <input placeholder="city,country code" onChange={this.onCityChange} value={this.state.selectedCity} />
```

File: src/App.js (hunk 2 above)
Comment (Sharakai) [reply src/App.js]: You're also missing a `key` on this `<div>`.
You **don't** need keys on the inner `<div>`s, but you do on this one, as it's the root element in each element item (after the `.map`)
```suggestion
                <div key={city}>{city}
```

File: src/App.js [hunk 8]
```diff
@@ -10,13 +10,35 @@ class App extends Component {
...
 
+    onPrint = () => {
+        console.log(this.state.subscribedCities);
+    };
+
+    renderCityData = (city, data) => {
+        return data.list.map( datum => {
+            return (
+                <div>{city}
+                    <div key={city + 'Date'}>Date: {datum.dt_txt}</div>
```
Comment (Sharakai) [inline src/App.js]: Redundant `key` on all of these `<div>` elements

File: src/App.js (hunk 1 above)
Comment (Sharakai) [reply src/App.js]: See above for how this function can be changed to just accept a single `data` parameter, and be able to extract the city name from `data.city.name`

**Review by Sharakai — CHANGES_REQUESTED** [review]
There's a fair amount going on in this PR, and this is the base of the application so it needs to be a good foundation. Happy to approve after the review comments have been either fixed, discussed, or rejected.

You can either reply to comments in standard textual fashion, or use an emote (hover over the top right of the comment) to agree (👍) or disagree (👎). If you disagree, please provide a comment as to why.
Do **NOT** take my opinions verbatim, and **DO** form your own ideas / opinions on your code, and feel free to fight back on any/all of the points raised.

File: src/App.js (hunk 1 above)
Comment (mattydono) [reply src/App.js]: Absolutely agree with passing through a single parameter, I don't know why I didn't do it in the first place.

File: src/App.js (hunk 2 above)
Comment (mattydono) [reply src/App.js]: That `{city}` was being used as a header for the displayed data and np I thought each item had to have its own unique key.

When you say root element are you referring to the fact that it's the parent container of the rendered data?

**Review by Sharakai — APPROVED** [review]

**Pull request merged**